### PR TITLE
fix #2773 with new precise encode

### LINKED ISSUE #2773: Can't specify precise crate version if there are multiple versions
``` toml
# depcrate/Cargo.toml 
[package]
name = "depcrate"
version = "0.1.0"

[dependencies]
spin = "0.3.5"
```

``` toml
# maincrate/Cargo.toml
[package]
name = "maincrate"
version = "0.1.0"

[dependencies]
spin = "0.4.0"
depcrate = { path = "../depcrate" }
```

A `cargo build` will result in getting spin versions `0.3.5` and `0.4.2`. Now, because of a bug, I'd like to downgrade `0.4.2` to `0.4.0`. Let's try this:

```
$cargo update -p spin --precise 0.4.0
error: There are multiple `spin` packages in your project, and the specification `spin` is ambiguous.
Please re-run this command with `-p <spec>` where `<spec>` is one of the following:
  spin:0.4.2
  spin:0.3.5
```

Well ok, that makes sense. So let's specify `spin:0.4.2`:

```
$cargo update -p spin:0.4.2 --precise 0.4.0
    Updating registry `https://github.com/rust-lang/crates.io-index`
error: no matching package named `spin` found (required by `depcrate`)
location searched: registry https://github.com/rust-lang/crates.io-index
version required: = 0.3.5
versions found: 0.4.0
```

I thought I just said to only touch `0.4.2`?

When I manually edit Cargo.lock and change all occurences of `0.4.2` to `0.4.0`, it builds fine.

### PR DESCRIPTION
**Pull request by gibix**
Changed the precise encode from <pkg>=<precise> to <pkg>=<present_version>-><future_version> in order to check the correct requirements.

cc @lu-zero

### CODE REVIEW

File: src/cargo/sources/registry/index.rs [hunk 1]
```diff
@@ -179,13 +179,18 @@ impl<'cfg> RegistryIndex<'cfg> {
...
+        // will have a precise version listed of the form
+        // `<pkg>=<p_req>o-><f_req>` where `<pkg>` is the name of a crate on
+        // this source, `<p_req>` is the version installed and `<f_req> is the
         // version requested (argument to `--precise`).
         let summaries = summaries.filter(|s| match source_id.precise() {
             Some(p) if p.starts_with(&*dep.name()) && p[dep.name().len()..].starts_with('=') => {
-                let vers = &p[dep.name().len() + 1..];
-                s.version().to_string() == vers
+                let vers: Vec<&str> = p[dep.name().len() + 1..].split("->").collect();
+                    if dep.version_req().matches(&Version::parse(vers[0]).unwrap()) {
```
Comment (lu-zero) [inline src/cargo/sources/registry/index.rs]: The indentation look off and might be nicer to have an `else` maybe?

File: src/cargo/sources/registry/index.rs [hunk 2]
```diff
@@ -186,11 +186,11 @@ impl<'cfg> RegistryIndex<'cfg> {
         let summaries = summaries.filter(|s| match source_id.precise() {
             Some(p) if p.starts_with(&*dep.name()) && p[dep.name().len()..].starts_with('=') => {
                 let vers: Vec<&str> = p[dep.name().len() + 1..].split("->").collect();
-                    if dep.version_req().matches(&Version::parse(vers[0]).unwrap()) {
-                        return vers[1] == s.version().to_string()
-                    }
-
+                if dep.version_req().matches(&Version::parse(vers[0]).unwrap()) {
+                    return vers[1] == s.version().to_string()
+                } {
```
Comment (lu-zero) [inline src/cargo/sources/registry/index.rs]: `else`

File: src/cargo/sources/registry/index.rs [hunk 3]
```diff
@@ -179,13 +179,18 @@ impl<'cfg> RegistryIndex<'cfg> {
...
+        // `<pkg>=<p_req>o-><f_req>` where `<pkg>` is the name of a crate on
+        // this source, `<p_req>` is the version installed and `<f_req> is the
         // version requested (argument to `--precise`).
         let summaries = summaries.filter(|s| match source_id.precise() {
             Some(p) if p.starts_with(&*dep.name()) && p[dep.name().len()..].starts_with('=') => {
-                let vers = &p[dep.name().len() + 1..];
-                s.version().to_string() == vers
+                let vers: Vec<&str> = p[dep.name().len() + 1..].split("->").collect();
+                if dep.version_req().matches(&Version::parse(vers[0]).unwrap()) {
+                    return vers[1] == s.version().to_string()
```
Comment (lu-zero) [inline src/cargo/sources/registry/index.rs]: drop the `return` so rustfmt will be happy and squash the set in a single patch please.

File: src/cargo/sources/registry/index.rs [hunk 4]
```diff
@@ -179,13 +179,18 @@ impl<'cfg> RegistryIndex<'cfg> {
...
-        // `<pkg>` is the name of a crate on this source and `<req>` is the
+        // will have a precise version listed of the form
+        // `<pkg>=<p_req>o-><f_req>` where `<pkg>` is the name of a crate on
+        // this source, `<p_req>` is the version installed and `<f_req> is the
         // version requested (argument to `--precise`).
         let summaries = summaries.filter(|s| match source_id.precise() {
             Some(p) if p.starts_with(&*dep.name()) && p[dep.name().len()..].starts_with('=') => {
-                let vers = &p[dep.name().len() + 1..];
-                s.version().to_string() == vers
+                let vers: Vec<&str> = p[dep.name().len() + 1..].split("->").collect();
```
Comment (alexcrichton) [inline src/cargo/sources/registry/index.rs]: Could this use `splitn` with 2, and avoid the intermediate `Vec` allocation by using `next().unwrap()` as well?

**Comment by alexcrichton** [comment]
Nice! I think there may be a merge conflict now though?

**Comment by matklad** [comment]
ping @alexcrichton

**Pull request merged**